### ISSUE TITLE
search.largeFiles not respected by indexed search

### ISSUE THREAD

**Issue by dadlerj** [sourcegraph/sourcegraph#12443]
Reported by https://app.hubspot.com/contacts/2762526/company/693777200

The user set search.largeFiles in his site config, then added a series of globs (e.g. `*.ts`). I then watched him run a search for the term that should have returned a match in a >1mb generated .ts file. It did not show up with `index:only` in the query.

Unfortunately, when we tried `index:no`, the search timed out (even with `count:1000`, it took so long that we hit a 504 gateway timeout), so there's no easy way to test given the scale of their monorepo. 

I know that indexing has to re-run before it would be available, so I'll let them know to try waiting a few hours or a day with that in the config. Filing this preemptively in case that doesn't solve it.

**Comment by rvantonder** [comment]
Flagging as potentially related to `index:no` 504 things in #12347

**Comment by dadlerj** [comment]
Anything we can do to investigate this? Curious if any diagnostic data would be helpful here @rvantonder

**Comment by rvantonder** [comment]
Based on this sharing similar traits as the issue in  #12347 (which is: large text files on the unindexed search path), it may be the same root cause which is: syntect, the syntax highlighting service, can't deal with it. In comment https://github.com/sourcegraph/sourcegraph/issues/12347#issuecomment-669711123 I showed some of the logs that pop up on the frontend when this happens.

Knowing whether their instance reports in the frontend, something like:

```
frontend | ERROR syntax highlighting failed (this is a bug, please report it), filepath: <STUFF> error: http://localhost:9238: HSS worker timeout while serving request
```

(i.e., search for "HSS worker timeout), or anything reported by the syntect server like:

```
 syntect_server | 2020/08/06 05:24:48 worker 77: restarting due to timeout
```
would help confirm or eliminate this as the cause.

In general, if we can grab some logs (Frontend probably likely the most useful) when search for this file could be helpful.

**Comment by keegancsmith** [comment]
For visibility here is the next steps I posted. Replace `$REPO` with the problematic repo.

Our `searcher` service works by fetching the full working copy of a repo, even if your query is for a very specific file. This is normally fine, but my expectation of the issue here is we are just failing to fetch a copy of your `$REPO` repo (which I am guessing is quite large).

To try help confirm the issue can you look for mentions of `$REPO` in your gitserver and searcher logs.

This is just a guess, to help understand exactly what is going on here a trace would be useful. Can you add `?trace=1` to your search page, specify timeout:30s and get the trace via these instructions: https://docs.sourcegraph.com/admin/observability/tracing

They are slightly out of date, you can now access Jaeger at http://sourcegraph.example.com/-/debug/jaeger/search

I don't really have a great workaround after poking in our source code right now, so getting to the bottom of what exactly is happening is the next step. That should lead to us either making an adjustment to your cluster or shipping a code change.

**Comment by keegancsmith** [comment]
TL;DR issue is archive fetch timeouts. Bumped timeout from 2m to 10m. Fix will be in master soon and branch cut for the next release is tomorrow.

So the root cause here is not that largeFiles doesn't work, it is we are failing to fetch the monorepo. The logs provided from the customer had the following:

```
t=2020-08-12T15:28:07+0000 lvl=warn msg="Long exec request" repo=$REPO args="[archive --worktree-attributes --format=tar $COMMIT --]" duration=2m0.078s
```

2m is a very suspicious timeout. Indeed we have a hardcoded timeout https://github.com/sourcegraph/sourcegraph/blob/v3.18.0/internal/store/store.go#L105
I bumped it here https://github.com/sourcegraph/sourcegraph/pull/12969

Additionally I improved logging for both gitserver and searcher when this sort of issue comes up: https://github.com/sourcegraph/sourcegraph/pull/12968 and https://github.com/sourcegraph/sourcegraph/pull/12970

As a sidenote I requested the customer provide a jaeger trace. What a lovely smooth process (on my side at least). I just dragged the JSON into the Jaeger UI and got immediate insight into the problem.

**Comment by dadlerj** [comment]
@keegancsmith just so I understand, by lifting the timeout here, we allow custom user-specified `timeouts:` to work, but non-indexed searches could still take multiple minutes just to transfer the archive? Or is the idea that this is fixed through caching, as long as it can get transferred at least once? Curious what the new user experience willb e

**Comment by keegancsmith** [comment]
This timeout is unrelated to the `timeout:` in a query. It is a background fetch from gitserver into a local cache. I haven't made it unlimited, but instead 10m. I should maybe be more conservative and raise to say 30m. I prefer keeping a timeout due to things that get stuck.

Essentially it is the time it takes to run `git archive`. For most repos that is just a few seconds. However, in this case the repo seems particularly large. I haven't seen us fail before in this case, even with testing of large monorepos / some other customers. So this is a particularly large repo / network transfer speeds slowed it down.

So user experience can be broken down into two parts:
1. Cached - user experience is time it takes to search the monorepo. This may still be slow, but by the sounds of it there are lots of large artifacts in the repo we won't search, so may still be acceptable.
2. Uncached - We quickly (500ms?) return to user saying we are still fetching. If they specify timeout we wait the full timeout before returning to the user.

**Comment by keegancsmith** [comment]
I just looked into this. It looks like we respect the option. However, we will skip files that don't look like text files / have too many trigrams: https://sourcegraph.com/github.com/sourcegraph/zoekt@ae29571c327d4be63ba4cf277aa8f9e7089952bb/-/blob/build/builder.go?subtree=true#L360

My guess this is the case for the missing file. @dadler Do we have more information on what the missing file actually is? I am out of office the next two days, but given the above link someone should be able to make progress on this issue if needed.

**Comment by dadlerj** [comment]
Just renamed this issue to focus on the original concern. Let's focus on why zoekt isn't indexing large files. I know the files are large/generated .ts files, but I don't know more than that. Will try to pull in someone else from search to help here!

**Comment by rvantonder** [comment]
TL;DR I think Keegan's right, that check on the number of trigrams may cause a file to not be indexed. I confirmed it with a representative file.

What I did: took the example `ts` file contents [here](https://doc.qt.io/qtinstallerframework/qt-installer-framework-translations-packages-com-vendor-product-meta-de-ts.html) and duplicated it until I had a 4MB file (aside, this is XML markup, not typescript, not really important, but that's an example reference we were pointed to). I then did some random character substitutions in the file so that there are more unique trigrams (just using the duplicated contents yields a low number of trigrams, I wanted to see if it was plausible for a 4MB text file to have > 20K unique trigrams by doing this). And yes, it turns out it is, and the file did not get indexed. I reran the index command with a higher max trigram count:

`zoekt-git-index -max_trigram_count 40000 -file_limit 10000000 -branches master -prefix origin/ . `

And that succeeded. It turns out that there were around 30K unique trigrams in my random 4MB text file. So, we should set this number to somewhere in the region of 50K to 100K for that file, I would guess.

Also of note: The generated `.zoekt` file that fails to index has this text at the very beginning:

`NOT-INDEXED: number of trigrams exceeds 200001`. Maybe that would show up in `.zoekt` files on customer instances for skipped files like this one. But, I don't think we need to confirm this, we should just set the number higher.

I haven't figured out how to go about setting a higher `-max_trigram_count`. Do we change/expose the option in Sourcegraph's Zoekt code, or update our Zoekt fork with a higher default, or some other deploy config?

I'll avoid changing the title of this issue, but if this resolves, it would more accurately be "Max trigram count is not big enough for certain large text files, and may be skipped by indexing"

**Comment by keegancsmith** [comment]
> I haven't figured out how to go about setting a higher -max_trigram_count. Do we change/expose the option in Sourcegraph's Zoekt code, or update our Zoekt fork with a higher default, or some other deploy config?

We don't provide a nice way, but we can expose this in the same way we expose the large files options (via config or an in-repo file).

The alternative solution is that if a file matches the large files glob, we don't apply the max trigram check. That to me seems more ideal.

Note: there is a performance trade off, the more unique trigrams you have, the more memory zoekt uses. It has this max trigram check to remove strange files. In this case though we want to include the file.

Combing large file glob check + max trigram sounds like the right approach and easy to do. I'll send out a change and we can chat about it later. Then if it sounds good, we can merge in and ship a patch release or tell the customer a custom zoekt image to use.

Note: they would likely have to force zoekt to reindex. If the repo is often changing ,they can just rely on it getting reindexed automatically, but otherwise if the repo doesn't change zoekt won't realise it needs to reindex.

**Comment by keegancsmith** [comment]
Fix at https://github.com/sourcegraph/zoekt/pull/59

@dadlerj I assume once this is merged in we need to do a patch release. Alternatively we can get the customer to just update zoekt in k8s (less release engineering work then). Does that sound good?

**Comment by dadlerj** [comment]
:tada: thank you!

**Comment by keegancsmith** [comment]
I just merged the fix in, and while writing the changelog entry I realised that the change can lead to infinite loops since the indexer assumes files are valid utf8 (which the change above skipped). Will send out another change soon.
